behave/configuration.py

```diff
@@ -452,14 +487,80 @@ def read_configuration(path):
     return result
 
 
+def read_toml(path):
+    """Read configuration from pyproject.toml file.
+
+    Configuration should be stored inside the 'tool.behave' table.
+
+    See https://www.python.org/dev/peps/pep-0518/#tool-table
+    """
+    # pylint: disable=too-many-locals, too-many-branches
+    config = json.loads(json.dumps(toml.load(path)))  # simple dict
+    config = config['tool']
+    config_dir = os.path.dirname(path)
+    result = {}
+
+    for dest, action in decode_options(config):
+        raw = config["behave"][dest]
+        if action == "store":
+            result[dest] = str(raw)
+        elif action in ("store_true", "store_false"):
+            result[dest] = bool(raw)
+        elif action == "append":
+            if dest == "userdata_defines":
+                continue    # -- SKIP-CONFIGFILE: Command-line only option.
+            # toml has native arrays and quoted strings, so there's no
+            # need to split by newlines or strip values
+            result[dest] = raw
+        else:
+            raise ValueError('action "%s" not implemented' % action)
+    format_outfiles_coupling(result, config_dir)
+
+    # -- STEP: Special additional configuration sections.
+    # SCHEMA: config_section: data_name
+    special_config_section_map = {
+        "formatters": "more_formatters",
+        "userdata":   "userdata",
+    }
+    for section_name, data_name in special_config_section_map.items():
+        result[data_name] = {}
+        try:
+            result[data_name] = values_to_str(config["behave"][section_name])
+        except KeyError:
+            result[data_name] = {}
+
+    return result
+
+
+def read_configuration(path, verbose=False):
+    ext = path.split(".")[-1]
+    parsers = {
+        "ini": read_configparser,
+        "cfg": read_configparser,
+        "behaverc": read_configparser,
+    }
+
+    if _TOML_AVAILABLE:
+        parsers["toml"] = read_toml
+    parse_func = parsers.get(ext, None)
+    if not parse_func:
+        if verbose:
+            print('Unable to find a parser for "%s"' % path)
+        return {}
+    parsed = parse_func(path)
+
+    return parsed
+
+
 def config_filenames():
     paths = ["./", os.path.expanduser("~")]
     if sys.platform in ("cygwin", "win32") and "APPDATA" in os.environ:
         paths.append(os.path.join(os.environ["APPDATA"]))
 
     for path in reversed(paths):
-        for filename in reversed(
-                ("behave.ini", ".behaverc", "setup.cfg", "tox.ini")):
+        for filename in reversed((
+            "behave.ini", ".behaverc", "setup.cfg", "tox.ini", "pyproject.toml"
```

Comment:
Maybe we can preserve the alphabetical order
```suggestion
            "behave.ini", ".behaverc", "pyproject.toml", "setup.cfg", "tox.ini"
```

Reply:
Paolo, I believe adding the new configuration file to the end of the list we follow the principle of least surprise. Using alphabetical order just for the sake of aesthetics could have a dramatic impact. :thinking: 

**Example:** A developer that had added some config to `pyproject.toml` (just for fun, w/o having it working), having their productive setup in `setup.cfg` or `tox.ini`, may get a broken setup all of a sudden for no apparent reason.

Reply:
It makes sense to me. I hope you can merge this one now. 🙂